### PR TITLE
complete compatibility regression testing

### PR DESCRIPTION
**Pull request by mqliang**
## Description
* split compCheck.sh as two script so we can build and test separately
* redirect services' logs to a file for easy of debug
* complete and enable test during rolling upgrade/rollback
* fix some minor bugs.

## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [ ] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)
## Release Notes
If you have tagged this as either backward-incompat or release-notes,
you MUST add text here that you would like to see appear in release notes of the
next release.

If you have a series of commits adding or enabling a feature, then
add this section only in final commit that marks the feature completed.
Refer to earlier release notes to see examples of text

## Documentation
If you have introduced a new feature or configuration, please add it to the documentation as well.
See https://docs.pinot.apache.org/developers/developers-and-contributors/update-document

### CODE REVIEW

File: compatibility-verifier/compCheck.sh [hunk 1]
```diff
@@ -34,61 +34,45 @@
...
-
 # Given a component and directory, start that version of the specific component
 function startService() {
   serviceName=$1
   dirName=$2
   # Upon start, save the pid of the process for a component into a file in /tmp/{component}.pid, which is then used to stop it
   pushd "$dirName"/pinot-tools/target/pinot-tools-pkg/bin  || exit 1
   if [ "$serviceName" = "zookeeper" ]; then
-    sh -c 'echo $$ > $0/zookeeper.pid; exec ./pinot-admin.sh StartZookeeper' "${dirName}" &
+    sh -c 'rm -rf /tmp/zkdir'
```
Comment (mcvsubbu) [inline compatibility-verifier/compCheck.sh:57]: Instead of `/tmp` use the working dir here

File: compatibility-verifier/compCheck.sh [hunk 2]
```diff
@@ -34,61 +34,45 @@
...
-    sh -c 'echo $$ > $0/server.pid; exec ./pinot-admin.sh StartServer' "${dirName}" &
+    sh -c 'echo $$ > $0/server.pid; exec ./pinot-admin.sh StartServer > ${0}/server.log 2>&1' "${dirName}" &
   elif [ "$serviceName" = "kafka" ]; then
-    sh -c 'echo $$ > $0/kafka.pid; exec ./pinot-admin.sh StartKafka -zkAddress localhost:2181/kafka' "${dirName}" &
+    sh -c 'echo $$ > $0/kafka.pid; exec ./pinot-admin.sh StartKafka -zkAddress localhost:2181/kafka > ${0}/kafka.log 2>&1' "${dirName}" &
   fi
+
+  # sleep 60s to allow service to do some initialization:
+  # 1. controller depends on zookeeper, if not wait zookeeper to be ready, controller will crash.
+  # 2. broker depends on controller, if not wait controller to be ready, broker will crash.
```
Comment (mcvsubbu) [inline compatibility-verifier/compCheck.sh:71]: you should include these sleeps in the start services method. Do the following:
- start Zookeeper
- call waitForZkReady
- start controller
- call waitForControllerReady
- start server
- start broker
- call waitForClusterReady

For now, in each of the callForXXXReady methods, just put a sleep for 10s, and add a TODO. We will add a check for zk, controller, etc. maybe in java via yaml file or using shell commands later.

File: compatibility-verifier/compCheck.sh [hunk 3]
```diff
@@ -217,29 +150,87 @@ if [ "$(lsof -t -i:8097 -s TCP:LISTEN)" ] || [ "$(lsof -t -i:8098 -sTCP:LISTEN)"
   exit 1
 fi
 
-# Setup initial cluster with olderCommit and do rolling upgrade
+# Setup initial cluster with olderCommit
 startServices "$oldTargetDir"
-#$COMPAT_TESTER pre-controller-upgrade.yaml; if [ $? -ne 0 ]; then exit 1; fi
+
+$COMPAT_TESTER "offline-table-create-op.yaml" 1; if [ $? -ne 0 ]; then exit 1; fi
```
Comment (mcvsubbu) [inline compatibility-verifier/compCheck.sh:156]: Don't add these yet. My plan is to change the compat-tester to take a dir name and execute all yamls under the directory. For now, just leaving one call commented as a placeholder is enough.

**Review by mcvsubbu — COMMENTED** [review]
I think you have tried to unify the schemas by adding hoursSinceEpoch? That may be ok for now. we will revisit it later.

Are you able to run all the tests fine?

Overall looks ok to me.

File: compatibility-verifier/compCheck.sh (hunk 1 above)
Comment (mqliang) [reply compatibility-verifier/compCheck.sh]: done

File: compatibility-verifier/compCheck.sh (hunk 3 above)
Comment (mqliang) [reply compatibility-verifier/compCheck.sh]: done

File: compatibility-verifier/compCheck.sh [hunk 4]
```diff
@@ -105,13 +102,18 @@ function stopService() {
...
 
 # Starts a Pinot cluster given a specific target directory
 function startServices() {
   dirName=$1
   startService zookeeper "$dirName"
+  # Controller depends on zookeeper, if not wait zookeeper to be ready, controller will crash.
+  waitForZkReady
   startService controller "$dirName"
+  # Broker depends on controller, if not wait controller to be ready, broker will crash.
+  waitForControllerReady
```
[the remark this answers is not in the document]
Comment (mqliang) [reply compatibility-verifier/compCheck.sh:120]: done

**Pull request merged**